source/wp-content/themes/wporg-parent-2021/functions.php

```diff
@@ -106,6 +107,114 @@ function prevent_arrow_emoji( $content ) {
 	return preg_replace( '/([←↑→↓↔↕↖↗↘↙])/u', '\1&#65038;', $content );
 }
 
+/**
+ * Merge the child theme's theme.json into the parent theme.json.
+ *
+ * Pull the parent theme's values for array settings out and merge them by slug
+ * with the values in the child theme. This prevents values in the child theme
+ * from blowing away the parent theme's settings.
+ *
+ * Additional settings are merged correctly, since they're objects (and merged
+ * by key).
+ *
+ * See https://github.com/WordPress/gutenberg/issues/40557.
+ *
+ * @param WP_Theme_JSON_Data $theme_json Parsed child theme.json.
+ *
+ * @return WP_Theme_JSON_Data The updated theme.json settings.
+ */
+function merge_parent_child_theme_json( $theme_json ) {
+	// Nothing to merge if this is not a child theme.
+	if ( get_template_directory() === get_stylesheet_directory() ) {
+		return $theme_json;
+	}
+
+	// Build a new theme.json object.
+	$new_data = array(
+		'version' => 2,
+	);
+
+	$child_theme = $theme_json->get_data();
+
+	if ( ! empty( $child_theme['settings'] ) ) {
+		$parent_theme_json_file = get_template_directory() . '/theme.json';
+		$parent_theme_json_data = wp_json_file_decode( $parent_theme_json_file, array( 'associative' => true ) );
+		$parent_theme           = new \WP_Theme_JSON_Gutenberg( $parent_theme_json_data );
+
+		// Get base theme.json settings.
+		$theme_settings = $parent_theme->get_settings();
+		$settings       = $child_theme['settings'];
+
+		// Define the array values here, so they can be updated if the theme.json schema changes.
+		$color_keys = [ 'duotone', 'gradient', 'palette' ];
+		$typog_keys = [ 'fontFamilies', 'fontSizes' ];
+		$space_keys = [ 'spacingSizes' ];
```

Comment:
Ideally, we wouldn't have to define these keys but I think it's _good enough_ for what we need right now. 

Also in an ideal world, core/Gutenberg handles this logic.

Reply:
> Ideally, we wouldn't have to define these keys but I think it's good enough for what we need right now.

We could probably build up which properties are "array" by traversing the schema, but by defining them like this we also know that this set definitely all have `slug` properties (there is an array with no slug, spacing.units, and that is not handled here). Anyway, it seemed easier to hardcode these than try to build them dynamically. I think the schema has been fairly stable since it was versioned 2.

> Also in an ideal world, core/Gutenberg handles this logic.

Agreed, but there was enough back and forth on [the GB issue](https://github.com/WordPress/gutenberg/issues/40557) about whether this should be the case that I don't see it happening soon.